core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -371,57 +371,65 @@ class ContainerProxy(
-        if (job.action.limits.logs.asMegaBytes == 0.MB) {
-          Future.successful(Right(activation))
-        } else {
-          val start = tid.started(this, LoggingMarkers.INVOKER_COLLECT_LOGS, logLevel = InfoLevel)
-          collectLogs(tid, job.msg.user, activation, container, job.action)
-            .andThen {
-              case Success(_) => tid.finished(this, start)
-              case Failure(t) => tid.failed(this, start, s"reading logs failed: $t")
-            }
-            .map(logs => Right(activation.withLogs(logs)))
-            .recover {
-              case LogCollectingException(logs) =>
-                Left(ActivationLogReadingError(activation.withLogs(logs)))
-              case _ =>
-                Left(ActivationLogReadingError(activation.withLogs(ActivationLogs(Vector(Messages.logFailure)))))
+    val activationWithLogs: Future[Either[ActivationLogReadingError, (WhiskActivation, Boolean)]] = activation
+      .flatMap {
+        {
+          case (activation, initFailure) =>
+            // Skips log collection entirely, if the limit is set to 0
+            if (job.action.limits.logs.asMegaBytes == 0.MB) {
+              Future.successful(Right(activation -> initFailure))
+            } else {
+              val start = tid.started(this, LoggingMarkers.INVOKER_COLLECT_LOGS, logLevel = InfoLevel)
+              collectLogs(tid, job.msg.user, activation, container, job.action)
+                .andThen {
+                  case Success(_) => tid.finished(this, start)
+                  case Failure(t) => tid.failed(this, start, s"reading logs failed: $t")
+                }
+                .map(logs => Right(activation.withLogs(logs) -> initFailure))
+                .recover {
+                  case LogCollectingException(logs) =>
+                    Left(ActivationLogReadingError(activation.withLogs(logs)))
+                  case _ =>
+                    Left(ActivationLogReadingError(activation.withLogs(ActivationLogs(Vector(Messages.logFailure)))))
+                }
             }
         }
+
       }
 
     // Storing the record. Entirely asynchronous and not waited upon.
-    activationWithLogs.map(_.fold(_.activation, identity)).foreach(storeActivation(tid, _))
+    activationWithLogs.map(_.fold(_.activation, _._1)).foreach(storeActivation(tid, _))
 
     // Disambiguate activation errors and transform the Either into a failed/successful Future respectively.
     activationWithLogs.flatMap {
-      case Right(act) if !act.response.isSuccess => Future.failed(ActivationUnsuccessfulError(act))
-      case Left(error)                           => Future.failed(error)
-      case Right(act)                            => Future.successful(act)
+      //if non-successful, init failures should fail, and non-applicationErrors should also fail
+      case Right((act, initFailure)) if !act.response.isSuccess && (initFailure || !act.response.isApplicationError) =>
```

Comment:
To make this a little bit more straightforward and reduce the size of the change, does it make sense to change https://github.com/apache/incubator-openwhisk/blob/master/common/scala/src/main/scala/whisk/core/containerpool/Container.scala#L114 to be a `containerError`? That way, I think all init failures are non-application errors and thus you can just check for `isSuccess | isApplicationError` here vs. having to thread the extra boolean flag through.

WDYT?

Reply:
Sounds reasonable! 👍